### PR TITLE
Add file type: .topostats

### LINKED ISSUE #51: [feature] : Add topostats file format support
### Is your feature request related to a problem?

Migrate the `.topostats` (HDF5 format) file format reader from TopoStats to AFMReader. This is the initial proposal for an open & interoperable AFM file format in HDF5 standard for AFM images. It will be iterated upon and likely re-designed but we need something to start with.

### Describe the solution you would like.

Migrate the `.topostats` file format to AFMReader as-is.

### Describe the alternatives you have considered.

N/A

### Sample Image

N/A

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by SylviaWhittle**
Closes #51 

This PR migrates support for the `.topostats` file format from [TopoStats](https://github.com/AFM-SPM/TopoStats) as part of our initiative to migrate all of TopoStats' `io.py` over. (#22 )

A function for reading hdf5 files and dumping the data into a dictionary is the main logic behind loading these files. Several tests have been added for it.

### CODE REVIEW

File: examples/example_01.ipynb
Comment (ns-rse) [inline examples/example_01.ipynb:1]: Not a big issue and certainly not something to do here but we could move the comments in the cells out to Markdown cells.

File: AFMReader/topostats.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    """
+    logger.info(f"Loading image from : {file_path}")
+    file_path = Path(file_path)
+    filename = file_path.stem
+    try:
+        with h5py.File(file_path, "r") as f:
+            data = unpack_hdf5(open_hdf5_file=f, group_path="/")
+
+            file_version = data["topostats_file_version"]
+            logger.info(f"TopoStats file version: {file_version}")
```
Comment (ns-rse) [inline AFMReader/topostats.py:46]: Perhaps include `filename` as is done in the `logger.error()` below so people know what file is which version?

```suggestion
            logger.info(f"[{filename}] TopoStats file version : {file_version}")
```

File: tests/test_topostats.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+BASE_DIR = Path.cwd()
+RESOURCES = BASE_DIR / "tests" / "resources"
+
+
+@pytest.mark.parametrize(
+    ("file_name", "topostats_file_version", "image_shape", "pixel_to_nm_scaling", "data_keys", "image_sum"),
+    [
+        pytest.param(
+            "sample_0.topostats",
+            0.1,
```
Comment (ns-rse) [inline tests/test_topostats.py:19]: Not for this PR but as we move forward with developing and standardising the `.topostats` file format and defining versions we should be looking to document precisely what each file version corresponds to with a detailed technical specification. A useful article on this is [A practical guide to writing technical specs - Stack Overflow](https://stackoverflow.blog/2020/04/06/a-practical-guide-to-writing-technical-specs/)

File: tests/test_topostats.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+BASE_DIR = Path.cwd()
+RESOURCES = BASE_DIR / "tests" / "resources"
+
+
+@pytest.mark.parametrize(
+    ("file_name", "topostats_file_version", "image_shape", "pixel_to_nm_scaling", "data_keys", "image_sum"),
+    [
+        pytest.param(
+            "sample_0.topostats",
```
Comment (ns-rse) [inline tests/test_topostats.py:18]: Perhaps pedantic but to make it easier to read/understand what's what in the `tests/resources/` directory we could use a nomenclature that reflects the version.

```suggestion
            "sample_0_1.topostats",
```

...would indicate the version `0.1` (would also require renaming `sample_0.topostats` > `sample_0_1.topostats`).

**Review by ns-rse — CHANGES_REQUESTED** [review]
Thanks for the comprehensive work on this @SylviaWhittle great to see all the tests and in-line examples as well as the notebook.

Some minor comments, not all for addressing in this Pull Request, have tried to make suggestions where things could be changed though.

**Comment by ns-rse** [comment]
Thanks for testing this @MaxGamill-Sheffield 

> On testing and reviewing this, I saw that there was no space for the .topostats files in the example_0.ipyb so I have added that.

You commit #0fdb9cc18 introduced a number of unintended changes to the notebook, the cell `execution_count` was now non-zero. I think this is because you haven't enabled `pre-commit` on this repository. In this instance its not a problem as `pre-commit.ci` is able to run the `nbstripout` hook and commit the changes to this PR directly but if you ever do additional work on this repository it would be worth enabling `pre-commit` locally so that all the hooks run prior to commits.

```bash
pre-commit install
```

...should do the trick.

**Comment by SylviaWhittle** [comment]
I think I've addressed the comments? Let me know if I've forgotten something.

I think the documentation on the format could be done in another PR? Or should I add it to this one?

**Comment by ns-rse** [comment]
> I think the documentation on the format could be done in another PR? Or should I add it to this one?

Definitely a separate issue and one that would benefit from wider discussion as to what people actually want to have in `.topostats` files. Garnering input from the wider community (beyond the Pyne lab) would be really useful in planing the formats specification.

**Pull request merged**